### PR TITLE
[Issue 4084] Create acceptance criteria dataset

### LINKED ISSUE #4084: [ETL:Export] Extend export output to include deliverable acceptance criteria
### Summary

This task is to be done after the roadmap export step of the ETL workflow has been extended to include deliverable bodyContent which contains acceptance criteria (see #4083).

In the export step of the ETL workflow, the `GitHubProjectETL` class exposes the method `extract_and_transfrom_in_memory` which is responsible for getting deliverable data from GitHub and storing it in an internal data structure of type `GitHubIssues`. See etl/github.py.

This task is to extend the functionality and output of `extract_and_transfrom_in_memory` as follows:

1. extract acceptance criteria data from roadmap_json
2. store acceptance criteria data in a new data structure to encapsulate description and state (boolean, i.e. is the checkbox checked)
3. return the data to caller

This task also includes integration of a markup parser to do the heavy lifting of parsing acceptance criteria from the bodyContent data (see https://github.com/HHS/simpler-grants-gov/issues/4082).

### Acceptance criteria

- [ ] Deliverable acceptance criteria are available for each deliverable in instances of GitHubIssues or in another new data structure

### PR DESCRIPTION
**Pull request by DavidDudas-Intuitial**
## Summary
Fixes #4084 

### Time to review: __10 mins__

[WIP] Scaffolding for extracting acceptance criteria from GitHub bodyContent values.

WIP

### CODE REVIEW

**Comment by DavidDudas-Intuitial** [comment]
@widal001 @jcrichlake I think we can leverage existing `roadmap_json` in `extract_and_transform_in_memory()` by adding key/value pair `issue_bodycontent`. If the roadmap json is extended to contain body content for each deliverable, then the new `AcceptanceCriteriaDataset` class in this PR will encapsulate the parsing, and can be used by the follow on load step in the ETL workflow.

File: analytics/src/analytics/etl/github.py [hunk 1]
```diff
@@ -206,6 +207,13 @@ def extract_and_transform_in_memory(self) -> list[dict]:
             pillar_field=roadmap.pillar_field,
         )
```
Comment (jcrichlake) [inline analytics/src/analytics/etl/github.py:209]: I think the data needs to be bubbled up from inside of the `export_roadmap_data_to_object` function.  I tested this some yesterday and my working functions weren't populating when just using the roadmap_json. But I'll explore this more this morning

Comment (DavidDudas-Intuitial) [reply analytics/src/analytics/etl/github.py:209]: @jcrichlake I prefer to not parse the body content during export. Instead, just include it (unparsed) as a value in the roadmap json returned by `export_roadmap_data_to_object`. Then we do not need to change the return type of `export_roadmap_data_to_object`. That method will continue to return the roadmap json but the json will contain a new key/value pair (e.g. `issue_bodycontent`) which will be parsed later during the load step.

roadmap-data

**Comment by DavidDudas-Intuitial** [comment]
@jcrichlake see my latest [commit](https://github.com/HHS/simpler-grants-gov/pull/4156/commits/895d4d4acdf074d50636b0aa434c196493c15cfd) in which I add issue body to existing roadmap query and data structure without need for new query, transformation, validation, or wrapper.

**Comment by DavidDudas-Intuitial** [comment]
After latest commit, roadmap data now contains raw body content for each deliverable:

roadmap-data-with-acceptance-critera

As a next step, we need to add @jcrichlake's checkbox parsing logic to `AcceptanceCriteriaDataset._parse_body_content()`

**Comment by DavidDudas-Intuitial** [comment]
Latest [commit](https://github.com/HHS/simpler-grants-gov/pull/4156/commits/ce62c2430888ac94e87b17fd28d2e7a485e95fe1) writes acceptance criteria totals to data warehouse, completing the end-to-end journey of the data from GitHub.

File: analytics/src/analytics/datasets/acceptance_criteria.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+        # init counters
+        total_criteria = 0
+        total_done = 0
+
+        # define regex to capture markdown sections:
+        # - captures section titles that start with "### "
+        # - captures section content until the next "###", EOF, or multiple blank lines
+        section_regex = r"^###\s+(.*?)\n([\s\S]*?)(?=\n###|\Z|\n{2,})"
+        sections = re.findall(section_regex, bodycontent, re.MULTILINE)
```
Comment (widal001) [inline analytics/src/analytics/datasets/acceptance_criteria.py:121]: I think this regex will miss the required content in at least one, maybe two scenarios:

- A given section (e.g. Acceptance Criteria) has two successive line breaks before it gets to a new header (see the Acceptance criteria section in [this example](https://regexr.com/8d52b)) -- this is likely to happen in at least a few cases.
- Acceptance Criteria or Metrics is the last section in the issue content and there is one return before the end of string (see [this example](https://regexr.com/8d530)) -- This might not be a concern with the `strip()` call earlier but I'd love to have at least one test that specifically checks for this edge case.


Is there a reason we're not simply splitting on the headers like in [this function](https://github.com/HHS/grants-product-and-delivery/blob/358abdeebbf967869a55ed1c21f4e7d7ccf48910/automations/product_strategy_metrics/calculate_metrics.py#L23-L41)?

File: analytics/tests/datasets/test_acceptance_criteria.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    """Test get_totals with a valid issue URL."""
+    totals = acceptance_criteria_dataset.get_totals(
+        "org/repo/issues/123",
+        AcceptanceCriteriaType.MAIN,
+        AcceptanceCriteriaNestLevel.ALL,
+    )
+
+    assert isinstance(totals, AcceptanceCriteriaTotal)
+    assert totals.criteria == 6
+    assert totals.done == 3
```
Comment (widal001) [inline analytics/tests/datasets/test_acceptance_criteria.py:129]: I'm glad that we're testing for at least a few cases here!

Since so much work is being offloaded to the parser method, it might be good to pull that out to a utility function that is simply invoked by `self._parse_body_content()` That would allow us to test it separately and more easily add explicit test scenarios for known edge cases as they arise.

**Review by widal001 — COMMENTED** [review]
Smart approach! And everything runs end to end as expected (see screenshot)

I flagged one specific scenario that I think will be a problem with the current regex to parse the sections. And left a request about testing the parser function separately so we can mock out those different scenarios.

Screenshot 2025-03-14 at 11 21 07 AM

File: analytics/src/analytics/datasets/acceptance_criteria.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+        # iterate sections
+        for section_name, section_body in sections:
+
+            # determine whether this section should be included in totals
+            if (
+                ac_type != AcceptanceCriteriaType.ALL
+                and section_name.lstrip("#").strip() != ac_type.value
+            ):
+                continue
```
Comment (widal001) [inline analytics/src/analytics/datasets/acceptance_criteria.py:138]: Oh also, this is probably fine for now, but the current code would count checkboxes in _any_ section (including those outside "Metrics" and "Acceptance criteria")

I _do think_ Lucas wants the count of checkboxes in both Acceptance Criteria and Metrics, but I'm not sure they want checkboxes outside those sections included (if they exist).

Maybe instead of a single enum value, we could check for a list of desired sections, like I have in [this function](https://github.com/HHS/grants-product-and-delivery/blob/358abdeebbf967869a55ed1c21f4e7d7ccf48910/automations/product_strategy_metrics/calculate_metrics.py#L44-L61). It could still be a list of enums.

Not a must have change right now, though, we just might want to anticipate that as potential feedback.

Comment (jcrichlake) [reply analytics/src/analytics/datasets/acceptance_criteria.py:138]: I would discuss that change, what if say we put in checkboxes for pre-requisites that need to be in place before we start something.  I don't think those checkboxes which would be needed for starting this story should count toward the progress of it

File: analytics/src/analytics/integrations/etldb/main.py [hunk 5]
```diff
@@ -124,13 +135,27 @@ def sync_data(dataset: EtlDataset, effective: str) -> None:
...
+        ac_total = (
+            ac.get_totals(
+                ghid,
+                AcceptanceCriteriaType.ALL,
+                AcceptanceCriteriaNestLevel.ALL,
+            )
+            if ac is not None
+            else AcceptanceCriteriaTotal()
+        )
+        result[ghid], _ = model.sync_deliverable(deliverable_df, ghid_map, ac_total)
```
Comment (widal001) [inline analytics/src/analytics/integrations/etldb/main.py:158]: @DavidDudas-Intuitial this looks like it's counting all sections.

File: analytics/src/analytics/datasets/acceptance_criteria.py (hunk 2 above)
Comment (widal001) [reply analytics/src/analytics/datasets/acceptance_criteria.py:121]: Is there a reason that we'd want to stop capture after hitting multiple new lines at all? I feel like just limiting it to either the next section or the end of the string should suffice, assuming scenario two isn't an issue with proper stripping.

File: analytics/src/analytics/datasets/acceptance_criteria.py (hunk 4 above)
Comment (DavidDudas-Intuitial) [reply analytics/src/analytics/datasets/acceptance_criteria.py:138]: You're right! Good catch. I'll fix it. The intention is too include BOTH sections not ALL sections

File: analytics/src/analytics/datasets/acceptance_criteria.py (hunk 2 above)
Comment (DavidDudas-Intuitial) [reply analytics/src/analytics/datasets/acceptance_criteria.py:121]: Good call. I'll iterate on this to improve it.

Comment (DavidDudas-Intuitial) [reply analytics/src/analytics/datasets/acceptance_criteria.py:121]: Thanks for pointing out the much simpler approach of splitting by section header.  I added it (see next commit) and it works great while simplifying the code.

File: analytics/src/analytics/datasets/acceptance_criteria.py (hunk 4 above)
Comment (DavidDudas-Intuitial) [reply analytics/src/analytics/datasets/acceptance_criteria.py:138]: Fixed in next commit.

File: analytics/src/analytics/integrations/etldb/main.py (hunk 5 above)
Comment (DavidDudas-Intuitial) [reply analytics/src/analytics/integrations/etldb/main.py:158]: Fixed

File: analytics/tests/datasets/test_acceptance_criteria.py (hunk 3 above)
Comment (DavidDudas-Intuitial) [reply analytics/tests/datasets/test_acceptance_criteria.py:130]: The `AcceptanceCriteriaDataset` class exists for the purpose of encapsulating the logic of parsing, so I prefer to keep the logic in the private method in the class instead of a utility function outside the class.

Comment (widal001) [reply analytics/tests/datasets/test_acceptance_criteria.py:130]: Fair enough, can we at least add more explicit tests for different scenarios?

I'd like to keep each test relatively focused on a specific edge case (e.g. checkboxes in sections other than "Acceptance Criteria" and "Metrics") so we are explicit about the desired behavior rather than just adding lots of different scenarios to a single mock input dataset.

**Review by widal001 — APPROVED** [review]
@DavidDudas-Intuitial thanks for addressing my core concerns!

I'd love to see us add a few more explicit test case scenarios around parsing to document the expected behavior around certain key edge cases, but don't let that hold up the merge.

Re-ran the pipeline and things looked good as far as I could tell!

Screenshot 2025-03-14 at 3 39 59 PM

**Comment by DavidDudas-Intuitial** [comment]
@widal001 I added 8 more tests for various edge cases

**Review by widal001 — APPROVED** [review]
LGTM! Thanks for adding some more tests
